pkg/util/debouncer/debouncer.go

```diff
@@ -70,229 +70,241 @@ func newMetrics(reg prometheus.Registerer, name string) *metrics {
+		return nil, errors.New("processHandler is required")
 	}
 
-	return &Debouncer[T]{
-		opts:         opts,
-		buffer:       make(chan T, opts.BufferSize),
-		debouncers:   make(map[string]*keyDebouncer[T]),
-		metrics:      newMetrics(opts.Reg, opts.Name),
-		errorHandler: opts.ErrorHandler,
+	if opts.ErrorHandler == nil {
+		opts.ErrorHandler = func(_ T, _ error) {}
 	}
+
+	return &Group[T]{
+		buffer:         make(chan T, opts.BufferSize),
+		debouncers:     make(map[T]*debouncer[T]),
+		processHandler: opts.ProcessHandler,
+		errorHandler:   opts.ErrorHandler,
+		minWait:        opts.MinWait,
+		maxWait:        opts.MaxWait,
+		metrics:        newMetrics(opts.Reg, opts.Name),
+	}, nil
 }
 
-func (d *Debouncer[T]) Add(value T) error {
+// Add will create a new debouncer for the given Key if it doesn't exist yet.
+// If a key has already a debouncer it will either reset the MinWait timer for
+// this key, or if they key is already running its process be no-op.
+func (g *Group[T]) Add(value T) error {
 	select {
-	case d.buffer <- value:
-		d.metrics.itemsAddedCounter.Inc()
+	case g.buffer <- value:
+		g.metrics.itemsAddedCounter.Inc()
 		return nil
 	default:
-		d.metrics.itemsDroppedCounter.Inc()
+		g.metrics.itemsDroppedCounter.Inc()
 		return ErrBufferFull
 	}
 }
 
-func (d *Debouncer[T]) Start(ctx context.Context, processFunc ProcessFunc[T]) {
-	d.ctx, d.cancel = context.WithCancel(ctx)
-	d.wg.Add(1)
+func (g *Group[T]) Start(ctx context.Context) {
+	g.ctx, g.cancel = context.WithCancel(ctx)
+	g.wg.Add(1)
 	go func() {
-		defer d.wg.Done()
+		defer g.wg.Done()
 		for {
 			select {
-			case <-d.ctx.Done():
+			case <-g.ctx.Done():
 				return
-			case value := <-d.buffer:
-				d.processValue(value, processFunc)
+			case value := <-g.buffer:
+				g.processValue(value)
 			}
 		}
 	}()
 }
 
-func (d *Debouncer[T]) Stop() {
-	if d.cancel != nil {
-		d.cancel()
-		d.wg.Wait()
+func (g *Group[T]) Stop() {
+	if g.cancel != nil {
+		g.cancel()
+		g.wg.Wait()
 	}
 }
 
-func (d *Debouncer[T]) SetMinWait(minWait time.Duration) {
-	d.opts.MinWait = minWait
-}
-
-func (d *Debouncer[T]) SetMaxWait(maxWait time.Duration) {
-	d.opts.MaxWait = maxWait
-}
-
-func (d *Debouncer[T]) processValue(value T, processFunc ProcessFunc[T]) {
-	key := d.opts.KeyFunc(value)
-	d.debouncersMu.Lock()
-	debouncer, ok := d.debouncers[key]
+func (g *Group[T]) processValue(key T) {
+	g.debouncersMu.Lock()
+	deb, ok := g.debouncers[key]
 	if !ok {
-		debouncer = newKeyDebouncer[T](d.opts.MinWait, d.opts.MaxWait)
-		d.debouncers[key] = debouncer
-	}
-	d.debouncersMu.Unlock()
+		deb = newDebouncer[T](g.minWait, g.maxWait, key, func(v T) {
+			g.processWithMetrics(g.ctx, v, g.processHandler)
 
-	wrappedProcessFunc := func(v T) {
-		d.processWithMetrics(d.ctx, v, processFunc)
-
-		d.debouncersMu.Lock()
-		defer d.debouncersMu.Unlock()
-		if current, exists := d.debouncers[key]; exists && current == debouncer {
-			delete(d.debouncers, key)
-		}
+			g.debouncersMu.Lock()
```

Comment:
We might be trying to lock an already locked mutex here. I did read the documentation for Lock but just want to make sure this is in fact what we want here.

Reply:
Yeah that's fine 👍